pkg/manifest/objects.go

```diff
@@ -0,0 +1,294 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	"istio.io/pkg/log"
+)
+
+// Object is an in-memory representation of a k8s object, used for moving between different representations
+// (Unstructured, JSON, YAML) with cached rendering.
+type Object struct {
+	object *unstructured.Unstructured
+
+	Group     string
+	Kind      string
+	Name      string
+	Namespace string
+
+	json []byte
+	yaml []byte
+}
+
+// NewObject creates a new Object and returns a ptr to it.
+func NewObject(u *unstructured.Unstructured, json, yaml []byte) *Object {
+	o := &Object{
+		object: u,
+		json:   json,
+		yaml:   yaml,
+	}
+
+	gvk := u.GetObjectKind().GroupVersionKind()
+	o.Group = gvk.Group
+	o.Kind = gvk.Kind
+	o.Name = u.GetName()
+	o.Namespace = u.GetNamespace()
+
+	return o
+}
+
+// Hash returns a unique, insecure hash based on kind, namespace and name.
+func Hash(kind, namespace, name string) string {
+	return strings.Join([]string{kind, namespace, name}, "/")
+}
+
+// ObjectsFromUnstructuredSlice returns an Objects ptr type from a slice of Unstructured.
+func ObjectsFromUnstructuredSlice(objs []*unstructured.Unstructured) (*Objects, error) {
+	ret := &Objects{}
+	for _, o := range objs {
+		ret.Items = append(ret.Items, NewObject(o, nil, nil))
+	}
+	return ret, nil
+}
+
+// ParseJSONToObject parses JSON to an Object.
+func ParseJSONToObject(json []byte) (*Object, error) {
+	o, gvk, err := unstructured.UnstructuredJSONScheme.Decode(json, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing json into unstructured object: %v", err)
+	}
+
+	u, ok := o.(*unstructured.Unstructured)
+	if !ok {
+		return nil, fmt.Errorf("parsed unexpected type %T", o)
+	}
+
+	return &Object{
+		object: u,
+		Group:  gvk.Group,
+		Kind:   gvk.Kind,
+		Name:   u.GetName(),
+		json:   json,
+	}, nil
+}
+
+// UnstructuredObject exposes the raw object, primarily for testing
+func (o *Object) UnstructuredObject() *unstructured.Unstructured {
+	return o.object
+}
+
+// GroupKind returns the GroupKind for o.
+func (o *Object) GroupKind() schema.GroupKind {
+	return o.object.GroupVersionKind().GroupKind()
+}
+
+// GroupVersionKind returns the GroupVersionKind for o.
+func (o *Object) GroupVersionKind() schema.GroupVersionKind {
+	return o.object.GroupVersionKind()
+}
+
+// Hash returns a unique hash for o.
+func (o *Object) Hash() string {
+	return Hash(o.Kind, o.Namespace, o.Name)
+}
+
+// JSON returns a JSON representation of o, using an internal cache.
+func (o *Object) JSON() ([]byte, error) {
+	if o.json != nil {
+		return o.json, nil
+	}
+
+	b, err := o.object.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	o.json = b
+	return b, nil
+}
+
+// YAML returns a YAML representation of o, using an internal cache.
+func (o *Object) YAML() ([]byte, error) {
+	if o.yaml != nil {
+		return o.yaml, nil
+	}
+	// TODO: there has to be a better way.
+	oj, err := o.JSON()
+	if err != nil {
+		return nil, err
+	}
+	o.json = oj
```

Comment:
This line is not necessary

Reply:
Are you sure? :-) It looks necessary but I may be missing something.

Reply:
@ostromart I think when you call o.JSON(), it already do this for you at line131?

Reply:
Yes, good point. Done.